pytential/symbolic/pde/reduce_fmms.py

```diff
@@ -0,0 +1,405 @@
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+from sumpy.kernel import (AxisTargetDerivative, AxisSourceDerivative,
+        KernelWrapper)
+
+from pymbolic.interop.sympy import PymbolicToSympyMapper, SympyToPymbolicMapper
+from pymbolic.mapper import Mapper
+from pymbolic.primitives import Product
+import sympy
+import functools
+from collections import defaultdict
+
+
+__all__ = (
+    "reduce_number_of_fmms",
+    )
+
+
+def reduce_number_of_fmms(int_gs, source_dependent_variables):
+    """
+    Reduce the number of FMMs needed for a system of
+    :class:`~pytential.symbolic.primitives.IntG` objects.
+
+    This is done by converting the ``IntG`` object to a matrix of polynomials
+    with d variables corresponding to d dimensions, where each variable represents
+    a (target) derivative operator along one of the axes. All the properties of
+    derivative operator that we want are reflected in the properties of the
+    polynomial including addition, multiplication and exact polynomial division.
+
+    This matrix is factored into two matrices, where the left hand side matrix
+    represents a transformation at the target, and the right hand side matrix
+    represents a transformation at the source.
+
+    If the expressions given are not linear, then the input expressions are
+    returned as is.
+
+    :arg int_gs: list of ``IntG`` objects.
+
+    :arg source_dependent_variables: list of :class:`pymbolic.primtives.Expression`
+        objects. When reducing FMMs, consider only these variables as dependent
+        on source. For eg: densities, source derivative vectors.
+
+    Note: there is no argument for target-dependent variables as the algorithm
+    assumes that there are no target-dependent-variables passed to this function.
+    (where a "target-dependent variable" represents a function discretized on the
+    targets)
+    """
+
+    dim = int_gs[0].target_kernel.dim
+    axis_vars = sympy.symbols(f"_x0:{dim}")
+
+    # A high level driver for this function should send int_gs that are common.
+    assert _check_int_gs_common(int_gs)
+
+    try:
+        mat, source_exprs = _create_matrix(int_gs, source_dependent_variables,
+            axis_vars)
+    except ValueError:
+        return int_gs
+
+    mat = sympy.Matrix(mat)
+    # Factor the matrix into two
+    try:
+        lhs_mat = _factor_left(mat, axis_vars)
+        rhs_mat = _factor_right(mat, lhs_mat)
+    except ValueError:
+        return int_gs
+
+    # If there are n inputs and m outputs,
+    #
+    #  - matrix = R^{m x n},
+    #  - LHS = R^{m x k},
+    #  - RHS = R^{k x n}.
+    #
+    # If k is not greater than or equal to n we are gaining nothing.
+    # Return as is.
+    if rhs_mat.shape[0] >= mat.shape[0]:
+        return int_gs
+
+    # Create SymPy Polynomial objects
```

Comment:
```suggestion
    # Cast to matrices of SymPy Polynomial objects
```

Reply:
Fixed in d87a0ba21fda883687707e9920bc07856ade619d